### DISCUSSION TITLE
Ordering question when DecodeRequest / EncodeResponse introduced

### DISCUSSION THREAD

**Discussion by wonderflow** [mosn/htnn#747] [Q&A]
I have read the whole doc: https://github.com/mosn/htnn/blob/main/site/content/en/docs/developer-guide/plugin_development.md

The ordering when DecodeRequest / EncodeResponse is not quite clear to me. Some questions:

Let's say we have three plugins in the execution order A->B->C. While B is the plugin which return the "waitAllData=true".  

1. In this case, plugin C's "DecodeHeader" will be executed after the the "DecodeRequest" of plugin B, this is not consistent when we don't use "waitAllData=true". Why can't we align the approach always finish executing all "DecodeHeader"?
2. Will the "DecodeData" of plugin B still be invoked if there's any logic inside?
3. If we have "EncodeResponse" in plugin B, does it mean "EncodeHeaders" of plugin A will also be invoked after "EncodeResponse" of plugin B?

**Comment by spacewander** [comment]
> In this case, plugin C's "DecodeHeader" will be executed after the the "DecodeRequest" of plugin B, this is not consistent when we don't use "waitAllData=true". Why can't we align the approach always finish executing all "DecodeHeader"?

It's by design. When people use a buffered plugin B, they may expect that if they change the header according to the body (for example, update the Content-Type), the modification of header is visible for the plugins after B.
If we execute all DecodeHeader first, then DecodeHeader of plugin C will be executed before the body handling of plugin B. It doesn't satisfy the requirement.

> Will the "DecodeData" of plugin B still be invoked if there's any logic inside?

No.

> If we have "EncodeResponse" in plugin B, does it mean "EncodeHeaders" of plugin A will also be invoked after "EncodeResponse" of plugin B?

Yes.

**Reply by spacewander** [reply]
> Will the "DecodeData" of plugin B still be invoked if there's any logic inside?

I need to correct my previous reply. The right answer is "it depends". If the DecodeHeader of B doesn't return WaitAllData, then DecodeData will be run instead.

**Reply by wonderflow** [reply]
The corner case is the last batch of the request body come in, the streamEnd=true and the DecodeHeader will not return "WaitAllDate=true", in this situation, what will happen? Maybe both "DecodeData" and "DecodeRequest" will run?
